gapic/samplegen/samplegen.py

```diff
@@ -436,48 +451,79 @@ def validate_expression(self, exp: str) -> wrappers.Field:
         """
         # TODO: handle mapping attributes, i.e. {}
         # TODO: Add resource name handling, i.e. %
-        indexed_exp_re = re.compile(
-            r"^(?P<attr_name>\$?\w+)(?:\[(?P<index>\d+)\])?$")
-
-        toks = exp.split(".")
-        match = indexed_exp_re.match(toks[0])
-        if not match:
-            raise BadAttributeLookup(
-                f"Badly formatted attribute expression: {exp}")
-
-        base_tok, previous_was_indexed = (match.groupdict()["attr_name"],
-                                          bool(match.groupdict()["index"]))
-        base = self.var_field(base_tok)
-        if not base:
-            raise UndefinedVariableReference(
-                "Reference to undefined variable: {}".format(base_tok))
-        if previous_was_indexed and not base.repeated:
-            raise BadAttributeLookup(
-                "Cannot index non-repeated attribute: {}".format(base_tok))
-
-        for tok in toks[1:]:
-            match = indexed_exp_re.match(tok)
+        chain_link_re = re.compile(
+            r"""
+            (?P<attr_name>\$?\w+)(?:\[(?P<index>\d+)\]|\{["'](?P<key>[^"']+)["']\})?$
```

Comment:
This works like magic.

Reply:
Is that good or bad? It's far from the worst regex I've ever written :P